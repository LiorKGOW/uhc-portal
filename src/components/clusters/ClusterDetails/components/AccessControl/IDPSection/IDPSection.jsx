import React from 'react';
import PropTypes from 'prop-types';
import get from 'lodash/get';

import {
  Card, Title, Button, CardBody, CardHeader, CardFooter,
} from '@patternfly/react-core';
import {
  Table,
  TableHeader,
  TableBody,
  TableVariant,
  cellWidth,
} from '@patternfly/react-table';
import {
  Skeleton,
} from '@redhat-cloud-services/frontend-components';
import ClipboardCopyLinkButton from '../../../../../common/ClipboardCopyLinkButton';

import links from '../../../../../../common/installLinks';
import { IDPTypeNames, getOauthCallbackURL, IDPNeedsOAuthURL } from '../../IdentityProvidersModal/IdentityProvidersHelper';

function IDPSection({
  clusterID, clusterConsoleURL, identityProviders, openModal,
}) {
  const columns = [
    { title: 'Name', transforms: [cellWidth(30)] },
    { title: 'Type', transforms: [cellWidth(30)] },
    { title: 'Auth Callback URL', transforms: [cellWidth(30)] },
  ];

  const actions = [
    {
      title: 'Delete',
      onClick: (_, __, rowData) => openModal('delete-idp', {
        clusterID,
        idpID:
        rowData.idpID,
        idpName: rowData.name.title,
        idpType: rowData.type.title,
      }),
    },
  ];

  const idpRow = idp => ({
    cells: [
      idp.name,
      get(IDPTypeNames, idp.type, idp.type),
      {
        title: IDPNeedsOAuthURL(idp.type) ? (
          <ClipboardCopyLinkButton className="idp-table-copy" text={getOauthCallbackURL(clusterConsoleURL, idp.name)}>
            Copy URL to clipboard
          </ClipboardCopyLinkButton>
        ) : 'N/A',
      },
    ],
    idpID: idp.id,
  });

  const learnMoreLink = <a rel="noopener noreferrer" href={links.UNDERSTANDING_IDENTITY_PROVIDER}>Learn more.</a>;

  const pending = !identityProviders.fulfilled && !identityProviders.error;

  const hasIDPs = !!identityProviders.clusterIDPList.length;

  return (
    pending ? (
      <Card>
        <CardHeader>
          <Skeleton size="md" />
        </CardHeader>
        <CardBody>
          <Skeleton size="lg" />
        </CardBody>
        <CardFooter>
          <Skeleton size="md" />
        </CardFooter>
      </Card>
    ) : (
      <Card>
        <CardBody>
          <Title size="lg" headingLevel="h3">
            Identity Providers
          </Title>
          <p>
            Configure identity providers to allow users to log into the cluster.
            {' '}
            {learnMoreLink}
          </p>
          { hasIDPs && (
            <Table
              aria-label="Identity Providers"
              actions={actions}
              variant={TableVariant.compact}
              cells={columns}
              rows={identityProviders.clusterIDPList.map(idpRow)}
            >
              <TableHeader />
              <TableBody />
            </Table>
          )}
          <Button onClick={() => openModal('create-identity-provider')} variant="secondary" className="add-idp-button">
            Add identity provider
          </Button>
        </CardBody>
      </Card>
    )
  );
}

IDPSection.propTypes = {
  clusterID: PropTypes.string.isRequired,
  clusterConsoleURL: PropTypes.string.isRequired,
  identityProviders: PropTypes.object.isRequired,
  openModal: PropTypes.func.isRequired,
};

export default IDPSection;
