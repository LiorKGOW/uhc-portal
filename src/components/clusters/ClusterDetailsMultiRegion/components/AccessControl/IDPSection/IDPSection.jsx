import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';

import {
  Card,
  CardBody,
  CardFooter,
  CardTitle,
  Dropdown,
  DropdownItem,
  DropdownList,
  MenuToggle,
  Stack,
  StackItem,
  Title,
  Tooltip,
} from '@patternfly/react-core';
import {
  ActionsColumn,
  Table,
  TableVariant,
  Tbody,
  Td,
  Th,
  Thead,
  Tr,
} from '@patternfly/react-table';
import Skeleton from '@redhat-cloud-services/frontend-components/Skeleton';

import { useNavigate } from '~/common/routing';
import { useFetchClusterIdentityProviders } from '~/queries/ClusterDetailsQueries/useFetchClusterIdentityProviders';

import links from '../../../../../../common/installLinks.mjs';
import ClipboardCopyLinkButton from '../../../../../common/ClipboardCopyLinkButton';
import { modalActions } from '../../../../../common/Modal/ModalActions';
import {
  getOauthCallbackURL,
  IDPformValues,
  IDPNeedsOAuthURL,
  IDPTypeNames,
} from '../../IdentityProvidersPage/IdentityProvidersHelper';

const IDPSection = (props) => {
  const {
    clusterID,
    clusterUrls,
    idpActions = {},
    clusterHibernating,
    isReadOnly,
    isHypershift,
    subscriptionID,
    cluster,
  } = props;
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const [dropdownOpen, setDropdownOpen] = React.useState(false);

  const region = cluster?.subscription?.xcm_id;

  const {
    clusterIdentityProviders: identityProvidersData,
    isLoading: isIdentityProvidersLoading,
    isError: isIdentityProvidersError,
  } = useFetchClusterIdentityProviders(clusterID, region);

  const identityProviders = identityProvidersData?.items;

  const learnMoreLink = (
    <a rel="noopener noreferrer" href={links.UNDERSTANDING_IDENTITY_PROVIDER} target="_blank">
      Learn more.
    </a>
  );

  const pending = isIdentityProvidersLoading && !isIdentityProvidersError;

  const hasIDPs = !!identityProviders?.length;

  const readOnlyReason = isReadOnly && 'This operation is not available during maintenance';
  const hibernatingReason =
    clusterHibernating && 'This operation is not available while cluster is hibernating';
  const notAllowedReason = (action) =>
    `You do not have permission to ${action} an identity provider. Only cluster owners, cluster editors, identity provider editors, and Organization Administrators can ${action} identity providers.`;
  const disableReason = readOnlyReason || hibernatingReason;
  const cannotCreateReason = disableReason || (!idpActions.create && notAllowedReason('add'));

  const IDPDropdownOptions = (
    <DropdownList>
      {Object.values(IDPTypeNames).map((idpName) => (
        <DropdownItem
          key={idpName}
          onClick={() =>
            navigate(`/details/s/${subscriptionID}/add-idp/${idpName.toLowerCase()}`, {
              state: {
                cluster,
                clusterIDPs: identityProviders,
                subscriptionID,
              },
            })
          }
        >
          {idpName}
        </DropdownItem>
      ))}
    </DropdownList>
  );

  const toggleRef = useRef();

  let addIDPDropdown = (
    <Dropdown
      isOpen={dropdownOpen}
      onOpenChange={(isOpen) => setDropdownOpen(isOpen)}
      toggle={{
        toggleRef,
        toggleNode: (
          <MenuToggle
            id="add-identity-provider"
            ref={toggleRef}
            isDisabled={cannotCreateReason}
            isExpanded={dropdownOpen}
            onClick={() => {
              setDropdownOpen(!dropdownOpen);
            }}
          >
            Add identity provider
          </MenuToggle>
        ),
      }}
    >
      {IDPDropdownOptions}
    </Dropdown>
  );
  if (cannotCreateReason) {
    addIDPDropdown = <Tooltip content={cannotCreateReason}>{addIDPDropdown}</Tooltip>;
  }

  const columnNames = {
    name: 'Name',
    type: 'Type',
    callbackUrl: 'Auth callback URL',
  };

  const idpActionResolver = (idp) => {
    const editIDPAction = {
      title: 'Edit',
      isAriaDisabled: !idpActions.update,
      onClick: () => {
        navigate(`/details/s/${subscriptionID}/edit-idp/${idp.name}`);
      },
    };
    const deleteIDPAction = {
      title: 'Delete',
      isAriaDisabled: !idpActions.delete,
      onClick: () => {
        dispatch(
          modalActions.openModal('delete-idp', {
            clusterID,
            idpID: idp.id,
            idpName: idp.name,
            idpType: IDPTypeNames[idp.type],
            region,
          }),
        );
      },
    };
    if (IDPTypeNames[idp.type] === IDPTypeNames[IDPformValues.HTPASSWD]) {
      return [deleteIDPAction];
    }
    return [editIDPAction, deleteIDPAction];
  };

  const idpRow = (idp) => {
    const actions = idpActionResolver(idp);
    return (
      <Tr key={idp.id}>
        <Td dataLabel={columnNames.name}>{idp.name}</Td>
        <Td dataLabel={columnNames.type}>{IDPTypeNames[idp.type] ?? idp.type}</Td>
        <Td dataLabel={columnNames.callbackUrl}>
          {IDPNeedsOAuthURL(idp.type) ? (
            <ClipboardCopyLinkButton
              className="access-control-tables-copy"
              text={getOauthCallbackURL(clusterUrls, idp.name, isHypershift)}
            >
              Copy URL to clipboard
            </ClipboardCopyLinkButton>
          ) : (
            'N/A'
          )}
        </Td>
        <Td isActionCell>
          <ActionsColumn items={actions} isDisabled={!!disableReason} />
        </Td>
      </Tr>
    );
  };

  return pending ? (
    <Card>
      <CardTitle>
        <Skeleton size="md" />
      </CardTitle>
      <CardBody>
        <Skeleton size="lg" />
      </CardBody>
      <CardFooter>
        <Skeleton size="md" />
      </CardFooter>
    </Card>
  ) : (
    <Card>
      <CardBody>
        <Stack hasGutter>
          <StackItem>
            <Title headingLevel="h2" size="lg" className="card-title">
              Identity providers
            </Title>
            <p>
              Configure identity providers to allow users to log into the cluster. {learnMoreLink}
            </p>
          </StackItem>
          <StackItem>{addIDPDropdown}</StackItem>
          <StackItem>
            {hasIDPs && idpActions.list && (
              <Table aria-label="Identity Providers" variant={TableVariant.compact}>
                <Thead>
                  <Tr>
                    <Th width={30}>{columnNames.name}</Th>
                    <Th width={30}>{columnNames.type}</Th>
                    <Th width={30}>{columnNames.callbackUrl}</Th>
                    <Th screenReaderText="Action" />
                  </Tr>
                </Thead>
                <Tbody>{identityProviders.map(idpRow)}</Tbody>
              </Table>
            )}
          </StackItem>
        </Stack>
      </CardBody>
    </Card>
  );
};

IDPSection.propTypes = {
  cluster: PropTypes.object,
  clusterID: PropTypes.string,
  clusterUrls: PropTypes.shape({
    console: PropTypes.string,
    api: PropTypes.string,
  }),
  idpActions: PropTypes.shape({
    get: PropTypes.bool,
    list: PropTypes.bool,
    create: PropTypes.bool,
    update: PropTypes.bool,
    delete: PropTypes.bool,
  }),
  clusterHibernating: PropTypes.bool,
  isReadOnly: PropTypes.bool,
  isHypershift: PropTypes.bool,
  subscriptionID: PropTypes.string,
};

export default IDPSection;
