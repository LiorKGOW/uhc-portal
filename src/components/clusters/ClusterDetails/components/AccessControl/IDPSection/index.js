import { connect } from 'react-redux';
import { getClusterIdentityProviders } from '../../IdentityProvidersModal/IdentityProvidersActions';
import { openModal } from '../../../../../common/Modal/ModalActions';

import IDPSection from './IDPSection';

const mapStateToProps = state => ({
  identityProviders: state.identityProviders.clusterIdentityProviders,
});

const mapDispatchToProps = {
  getClusterIdentityProviders,
  openModal,
};

export default connect(mapStateToProps, mapDispatchToProps)(IDPSection);
