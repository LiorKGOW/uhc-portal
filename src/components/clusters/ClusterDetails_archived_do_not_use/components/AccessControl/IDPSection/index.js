import { connect } from 'react-redux';

import { openModal } from '../../../../../common/Modal/ModalActions';
import { getClusterIdentityProviders } from '../../IdentityProvidersPage/IdentityProvidersActions';

import IDPSection from './IDPSection';

const mapStateToProps = (state) => ({
  identityProviders: state.identityProviders.clusterIdentityProviders,
  subscriptionID: state.clusters.details.cluster.subscription?.id,
});

const mapDispatchToProps = {
  getClusterIdentityProviders,
  openModal,
};

export default connect(mapStateToProps, mapDispatchToProps)(IDPSection);
