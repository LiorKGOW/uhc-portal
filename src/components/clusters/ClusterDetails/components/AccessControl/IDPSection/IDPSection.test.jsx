import React from 'react';
import { shallow } from 'enzyme';
import { Button } from '@patternfly/react-core';

import IDPSection from './IDPSection';

const baseIDPs = {
  clusterIDPList: [],
  pending: false,
  fulfilled: true,
  error: false,
};

describe('<IDPSection />', () => {
  it('should render (no IDPs)', () => {
    const openModal = jest.fn();
    const wrapper = shallow(<IDPSection
      canEdit={false}
      clusterID="fake id"
      identityProviders={baseIDPs}
      openModal={openModal}
      clusterConsoleURL="http://example.com/"
    />);
    expect(wrapper).toMatchSnapshot();
    wrapper.find(Button).simulate('click');
    expect(openModal).toBeCalledWith('create-identity-provider');
  });

  it('should render (IDPs pending)', () => {
    const IDPs = {
      ...baseIDPs,
      pending: true,
      fulfilled: false,
    };

    const wrapper = shallow(<IDPSection
      canEdit={false}
      clusterID="fake id"
      identityProviders={IDPs}
      openModal={jest.fn()}
      clusterConsoleURL="http://example.com/"
    />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render (with IDPs)', () => {
    const IDPs = {
      ...baseIDPs,
      clusterIDPList: [
        {
          name: 'hello',
          type: 'GithubIdentityProvider',
          id: 'id',
        },
        {
          name: 'hi',
          type: 'GoogleIdentityProvider',
          id: 'id',
        },
      ],
      fulfilled: true,
    };

    const openModal = jest.fn();
    const wrapper = shallow(<IDPSection
      canEdit={false}
      clusterID="fake id"
      identityProviders={IDPs}
      openModal={openModal}
      clusterConsoleURL="http://example.com/"
    />);
    expect(wrapper).toMatchSnapshot();
    wrapper.find(Button).simulate('click');
    expect(openModal).toBeCalledWith('create-identity-provider');
  });
});
