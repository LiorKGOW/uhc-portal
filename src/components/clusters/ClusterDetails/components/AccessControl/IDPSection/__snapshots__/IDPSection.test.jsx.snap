// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IDPSection /> should render (IDPs pending) 1`] = `
<Card>
  <CardHeader>
    <Skeleton
      isDark={false}
      size="md"
    />
  </CardHeader>
  <CardBody>
    <Skeleton
      isDark={false}
      size="lg"
    />
  </CardBody>
  <CardFooter>
    <Skeleton
      isDark={false}
      size="md"
    />
  </CardFooter>
</Card>
`;

exports[`<IDPSection /> should render (no IDPs) 1`] = `
<Card>
  <CardBody>
    <Title
      headingLevel="h3"
      size="lg"
    >
      Identity Providers
    </Title>
    <p>
      Configure identity providers to allow users to log into the cluster.
       
      <a
        href="https://docs.openshift.com/container-platform/4.2/authentication/understanding-identity-provider.html"
        rel="noopener noreferrer"
      >
        Learn more.
      </a>
    </p>
    <Component
      className="add-idp-button"
      onClick={[Function]}
      variant="secondary"
    >
      Add identity provider
    </Component>
  </CardBody>
</Card>
`;

exports[`<IDPSection /> should render (with IDPs) 1`] = `
<Card>
  <CardBody>
    <Title
      headingLevel="h3"
      size="lg"
    >
      Identity Providers
    </Title>
    <p>
      Configure identity providers to allow users to log into the cluster.
       
      <a
        href="https://docs.openshift.com/container-platform/4.2/authentication/understanding-identity-provider.html"
        rel="noopener noreferrer"
      >
        Learn more.
      </a>
    </p>
    <Component
      actions={
        Array [
          Object {
            "onClick": [Function],
            "title": "Delete",
          },
        ]
      }
      aria-label="Identity Providers"
      cells={
        Array [
          "Name",
          "Type",
        ]
      }
      rows={
        Array [
          Object {
            "cells": Array [
              "hello",
              "GitHub",
            ],
            "idpID": "id",
          },
          Object {
            "cells": Array [
              "hi",
              "Google",
            ],
            "idpID": "id",
          },
        ]
      }
      variant="compact"
    >
      <TableHeader />
      <TableBody />
    </Component>
    <Component
      className="add-idp-button"
      onClick={[Function]}
      variant="secondary"
    >
      Add identity provider
    </Component>
  </CardBody>
</Card>
`;
